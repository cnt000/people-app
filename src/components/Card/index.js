import React from 'react'
import styled, {css} from 'styled-components';

const Div = styled.div`
  padding: 0.25em 1em;
  width: 10em;
  height: 16em;
  border: 2px solid black;
  color: black;
  border-radius: 0.25em;
  background-color: lightgray;
  margin: 0.2em;
  font-size: 16px;

  ${props => props.showed && css`
    background: green;
  `}
`;

class Card extends React.Component {

  constructor(props) {
    super(props)
    this.state = { showed: this.props.showed }
  }

  show() {
    this.setState({ showed: true })
  }

  render() {
    return <Div 
      showed={this.state.showed}
      onClick={() => this.show()}>
        {this.state.showed && this.props.value}
      </Div>
  }
}

export default Card