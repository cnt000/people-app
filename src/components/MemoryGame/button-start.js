import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const Div = styled.div`
  margin: auto auto;
  text-align: center;
`

const ButtonStart = ({ onClick }) => {
  return (
    <Div onClick={onClick}>
      <h1>Memory Game</h1>
      <p>Click to start</p>
    </Div>
  )
}

ButtonStart.propTypes = {
  onClick: PropTypes.func.isRequired,
}

export default ButtonStart
