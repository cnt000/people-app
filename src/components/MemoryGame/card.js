import React from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'

const Div = styled.div`
  padding: 0.5em 0.4em;
  width: 10em;
  height: 14em;
  border: 2px solid black;
  color: black;
  border-radius: 0.25em;
  background: linear-gradient(
    210deg,
    rgba(26, 188, 156, 1) 0%,
    rgba(142, 68, 173, 1) 100%
  );
  margin: 0.4em 0.1em;
  font-size: 16px;

  ${props =>
    props.showed &&
    css`
      background: white;
    `};
`

const Card = ({ value, showed, onClick }) => {
  return (
    <Div onClick={onClick} showed={showed}>
      {showed && value}
    </Div>
  )
}

Card.propTypes = {
  value: PropTypes.string.isRequired,
  showed: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
}

export default Card
