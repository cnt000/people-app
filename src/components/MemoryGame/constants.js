export const LOAD_CARDS = 'people-app/MemoryGame/LOAD_CARDS';
export const SHOW_CARD = 'people-app/MemoryGame/SHOW_CARD';
export const CHECK_SELECTED_CARDS = 'CHECK_SELECTED_CARDS';
export const PONG = 'PONG';
export const DEFAULT_LOCALE = 'en';