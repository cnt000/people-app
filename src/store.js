import { createStore, applyMiddleware, compose, combineReducers } from 'redux'
import { routerMiddleware, routerReducer } from 'react-router-redux'
import createHistory from 'history/createBrowserHistory'
import { combineEpics, createEpicMiddleware } from 'redux-observable'
import {
  memoryGameShowCardEpic,
  memoryGameCleanSelectedEpic,
  memoryGameHideCoupleEpic,
  memoryGameTimerStartEpic
} from './components/MemoryGame/epic'
import { memoryGameReducer } from './components/MemoryGame/reducer'

export const history = createHistory()

const initialState = {}
const enhancers = []

const epicMiddleware = createEpicMiddleware()
const middleware = [routerMiddleware(history), epicMiddleware]

if (process.env.NODE_ENV === 'development') {
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__

  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension())
  }
}

const reducer = combineReducers({
  routerReducer,
  memoryGameReducer,
})

export const rootEpic = combineEpics(
  memoryGameShowCardEpic,
  memoryGameCleanSelectedEpic,
  memoryGameHideCoupleEpic,
  memoryGameTimerStartEpic,
)

const composedEnhancers = compose(
  applyMiddleware(...middleware),
  ...enhancers,
)

const store = createStore(reducer, initialState, composedEnhancers)

epicMiddleware.run(rootEpic)

export default store
