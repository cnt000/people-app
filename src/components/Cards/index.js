import React from "react"
import PropTypes from 'prop-types'
import styled, { keyframes } from "styled-components"
import Card from "../Card"

const gradient = keyframes`
  0% {
    background-position: 0% 50%
  }
  50% {
    background-position: 100% 50%
  }
  100% {
    background-position: 0% 50%
  }
`

const Div = styled.div`
  width: 90vw;
  height: 90vh;
  border: 1px solid black;
  background: linear-gradient(-45deg, #EE7752, #E73C7E, #23A6D5, #23D5AB);
  background-size: 170% 100%;
  animation: ${gradient} 15s ease infinite;
  border-radius: 0.25em;
  padding: .4em;
`

const Ul = styled.ul`
  display: grid;
  grid-template-columns: 25% 25% 25% 25%;
  grid-template-rows: 25% 25% 25% 25%;
  margin: 0;
  padding: 0;
`
const Li = styled.li`
  list-style: none;
  margin: 0;
  padding: 0;
`

const Cards = ({ cards, showCard }) => (
  <Div>
    <Ul>
      {cards && cards.map((card, position) =>
        <Li key={position}>
          <Card
            value={card.value}
            showed={card.showed}
            onClick={() => showCard(position)}
          />
        </Li>
      )}
    </Ul>
  </Div>
)

Cards.propTypes = {
  cards: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      showed: PropTypes.bool.isRequired
    }).isRequired
  ).isRequired
}

export default Cards
