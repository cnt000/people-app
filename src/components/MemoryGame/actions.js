import { LOAD_CARDS, SHOW_CARD, CHECK_COUPLE, START_GAME } from './constants'

export function loadCards() {
  return {
    type: LOAD_CARDS,
  }
}

export function showCard(cardPosition) {
  return {
    type: SHOW_CARD,
    cardPosition,
  }
}

export function startGame() {
  return {
    type: START_GAME,
  }
}

export function checkCouple() {
  return {
    type: CHECK_COUPLE,
  }
}
