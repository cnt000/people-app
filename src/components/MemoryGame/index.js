import { connect } from 'react-redux'
import Cards from './cards'
import { showCard, startGame, checkCouple } from './actions'
import {
  PLAYING_STATE,
  FINISHED_GAME_STATE,
  FIRST_OF_COUPLE_STATE,
  SECOND_OF_COUPLE_STATE,
  CORRECT_COUPLE_STATE,
  INCORRECT_COUPLE_STATE,
} from './constants'

const mapStateToProps = state => {
  return {
    cards: state.memoryGameReducer.cards,
    isPlaying:
      state.memoryGameReducer.gameState === PLAYING_STATE ||
      state.memoryGameReducer.gameState === FIRST_OF_COUPLE_STATE ||
      state.memoryGameReducer.gameState === SECOND_OF_COUPLE_STATE ||
      state.memoryGameReducer.gameState === CORRECT_COUPLE_STATE ||
      state.memoryGameReducer.gameState === INCORRECT_COUPLE_STATE,
    hasWin: state.memoryGameReducer.gameState === FINISHED_GAME_STATE,
    isClickable: state.memoryGameReducer.gameState === PLAYING_STATE ||
      state.memoryGameReducer.gameState === FIRST_OF_COUPLE_STATE,
    gameState: state.memoryGameReducer.gameState,
  }
}

const mapDispatchToProps = dispatch => ({
  showCard: id => dispatch(showCard(id)),
  startGame: () => dispatch(startGame()),
  checkCouple: gameState =>
    gameState === SECOND_OF_COUPLE_STATE ? dispatch(checkCouple()) : '',
})

const MemoryGame = connect(
  mapStateToProps,
  mapDispatchToProps
)(Cards)

export default MemoryGame
