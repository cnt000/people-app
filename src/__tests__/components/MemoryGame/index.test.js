import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import MemoryGame from '../../../components/MemoryGame'
import Cards from '../../../components/MemoryGame/cards'
import Card from '../../../components/MemoryGame/card'
import { timerGame } from '../../../modules/Helpers'
import store from '../../../store'

describe('MemoryGame', () => {
  it('should render (without snapshot)', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <Provider store={store}>
        <MemoryGame />
      </Provider>,
      div
    )
    ReactDOM.unmountComponentAtNode(div)
  })

  it('should render', () => {
    const output = shallow(
      <Provider store={store}>
        <MemoryGame />
      </Provider>
    )
    expect(shallowToJson(output)).toMatchSnapshot()
  })

  it('Before game start: isPlaying false', () => {
    const enzymeWrapper = mount(
      <Provider store={store}>
        <MemoryGame />
      </Provider>
    )
    expect(enzymeWrapper.find(Cards).prop('isPlaying')).toEqual(false)
  })

  it('Game started: click on start-button, isPlaying true', () => {
    const enzymeWrapper = mount(
      <Provider store={store}>
        <MemoryGame />
      </Provider>
    )

    enzymeWrapper.find('.start-button').simulate('click')
    expect(enzymeWrapper.find(Cards).prop('isPlaying')).toEqual(true)
  })

  it('Game started: first Card is showed', () => {
    const enzymeWrapper = mount(
      <Provider store={store}>
        <MemoryGame />
      </Provider>
    )

    expect(
      enzymeWrapper
        .find(Card)
        .first()
        .prop('showed')
    ).toEqual(true)
  })

  // it('wait', () => {
  //   const enzymeWrapper = mount(<Provider store={store}>
  //     <MemoryGame />
  //   </Provider>,)

  // jest.useFakeTimers(); // this must be called before any async things happen

  // setTimeout(() => {
  //   // Note the placement of this try/catch is important.
  //   // You'd think it could be placed most anywhere, but nope...
  //   try {
  //     // If this assertion fails, an err is thrown.
  //     // If we do not catch()...done.fail(e) it,
  //     // then this test will take jasmine.DEFAULT_TIMEOUT_INTERVAL (5s unless you override)
  //     // and then fail with an unhelpful message.
  //     expect(enzymeWrapper.find(Card).first().prop('showed').toEqual(false))
  //   } catch(e) {
  //   }
  // }, 11000);
  // jest.runTimersToTime(11000);
  //   //expect(enzymeWrapper.find(Card).first().prop('showed')).toEqual(false)
  // })
})
