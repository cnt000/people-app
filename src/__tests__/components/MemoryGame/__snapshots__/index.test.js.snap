// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemoryGame should render 1`] = `<Connect(Cards) />`;
